README.md
# Binlist

[![Hex.pm](https://img.shields.io/hexpm/v/binlist.svg)](https://hex.pm/packages/binlist)
[![Build Status](https://travis-ci.org/javierjulio/binlist-elixir.svg?branch=master)](https://travis-ci.org/javierjulio/binlist-elixir)

Binlist is an Elixir client library for [the Binlist.net service](https://binlist.net).

## Installation

First, add Binlist to your `mix.exs` dependencies:

```elixir
def deps do
  [ {:binlist, "~> 0.1.0"} ]
end
```

Then, update your dependencies:

```
$ mix deps.get
```

## Usage

Make a request, specifying the bin number as an integer or string:

```elixir
{:ok, bin} = Binlist.find(45717360)
```

And if successful, `bin` is a `%Binlist.Bin{}` struct:

```elixir
bin.scheme  # "visa"
bin.type    # "debit"
bin.prepaid # false
```

If an error occurs either of the following are returned:

```elixir
# For a 404:
{:error, "Bin not found"}

# For a 429:
{:error, "Request limit reached"}

# For any other 4xx or 5xx status:
{:error, "Unknown error", status}
```

## Contributing

This is my first Elixir library so if you notice something wrong, please open an issue or better yet submit a pull request.

Make a change and then run tests `mix test` to verify your changes. Note that the tests are integration based.

Binlist.net sometimes doesn't return a field for one bin but it does for another. Since the library uses structs, if you notice any fields missing please submit an issue along with the bin introducing the field.

## Misc

As this was my first Elixir library, I looked to other projects to see how they were configured so I knew how to distribute on Hex. I liked the code I read in [ex_twilio](https://github.com/danielberkompas/ex_twilio) so I used the same approach to parsing a response.


CHANGELOG.md
Changelog
=========

## 0.1.0 (2017-08-18)

Initial release.


test/binlist_test.exs
defmodule BinlistTest do

  use ExUnit.Case

  doctest Binlist

  test "finds a valid bin" do
    {:ok, bin} = Binlist.find(45717360)

    assert bin.scheme == "visa"
    assert bin.type == "debit"
    refute bin.prepaid
    assert bin.brand == "Visa/Dankort"
    assert bin.bank.city == "Hjørring"
    assert bin.bank.name == "Jyske Bank"
    assert bin.bank.phone == "+4589893300"
    assert bin.bank.url == "www.jyskebank.dk"
    assert bin.country.alpha2 == "DK"
    assert bin.country.currency == "DKK"
    assert bin.country.emoji == "🇩🇰"
    assert bin.country.latitude == 56
    assert bin.country.longitude == 10
    assert bin.country.name == "Denmark"
    assert bin.country.numeric == "208"
    assert bin.number.length == 16
    assert bin.number.luhn
  end

  test "returns an error if bin not found" do
    assert {:error, "Bin not found"} == Binlist.find(123456)
  end

end


test/binlist/parser_test.exs
defmodule Binlist.ParserTest do

  use ExUnit.Case

  import Binlist.Parser

  defmodule DummyUser do
    defstruct name: nil
  end

  doctest Binlist.Parser

  test "populates struct from given module with response data" do
    response = %{body: "{\"name\": \"Javier Julio\"}", status_code: 200}

    assert {:ok, %DummyUser{name: "Javier Julio"}} == parse(response, DummyUser)
  end

  test "returns 'Bin not found' for 404 status" do
    response = %{body: "", status_code: 404}

    assert {:error, "Bin not found"} == parse(response, nil)
  end

  test "returns 'Request limit reached' for 429 status" do
    response = %{body: "", status_code: 429}

    assert {:error, "Request limit reached"} = parse(response, nil)
  end

  test "returns 'Unknown error' and status for any other error" do
    response = %{body: "", status_code: 500}

    assert {:error, "Unknown error", 500} = parse(response, nil)
  end

end


mix.exs
defmodule Binlist.Mixfile do
  use Mix.Project

  def project do
    [
      app: :binlist,
      version: "0.1.0",
      elixir: "~> 1.4",
      start_permanent: Mix.env == :prod,
      deps: deps(),
      package: package()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger, :httpoison, :poison]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:httpoison, "~> 0.13"},
      {:poison, "~> 3.1"},
      {:ex_doc, ">= 0.0.0", only: :dev}
    ]
  end

  def docs do
    [
      readme: "README.md",
      main: Binlist
    ]
  end

  defp package do
    [
      description: "An Elixir client library for the Binlist.net service.",
      files: ["lib", "mix.exs", "README.md", "LICENSE*"],
      maintainers: ["Javier Julio"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/javierjulio/binlist-elixir"}
    ]
  end

end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :binlist, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:binlist, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/binlist/parser.ex
defmodule Binlist.Parser do

  @type success          :: {:ok, [map]}
  @type error            :: {:error, String.t}

  @spec parse(HTTPoison.Response.t, module) :: success | error
  def parse(response, module) do
    handle_errors response, fn(body) ->
      Poison.decode!(body, as: target(module))
    end
  end

  defp target(module) when is_atom(module), do: module.__struct__
  defp target(other), do: other

  defp handle_errors(response, fun) do
    case response do
      %{body: body, status_code: status} when status in [200, 201] ->
        {:ok, fun.(body)}

      %{body: _, status_code: 404} ->
        {:error, "Bin not found"}

      %{body: _, status_code: 429} ->
        {:error, "Request limit reached"}

      %{body: _, status_code: status} ->
        {:error, "Unknown error", status}
    end
  end

end


lib/binlist.ex
defmodule Binlist do

  @moduledoc """
  A client library for the Binlist.net service.
  """

  use HTTPoison.Base

  alias Binlist.{Bin, Parser}

  @doc """
  Returns a Binlist.Bin struct for the requested bin number.

  ## Examples

      iex> Binlist.find(45717360)
      {:ok,
      %Binlist.Bin{bank: %Binlist.Bank{city: "Hjørring", name: "Jyske Bank",
        phone: "+4589893300", url: "www.jyskebank.dk"}, brand: "Visa/Dankort",
        country: %Binlist.Country{alpha2: "DK", currency: "DKK", emoji: "🇩🇰",
        latitude: 56, longitude: 10, name: "Denmark", numeric: "208"},
        number: %Binlist.Number{length: 16, luhn: true}, prepaid: false,
        scheme: "visa", type: "debit"}}

      iex> Binlist.find(123456)
      {:error, "Bin not found"}

  """
  def find(bin) do
    "https://lookup.binlist.net/#{bin}"
    |> Binlist.get!(["Accept-Version": "3"])
    |> Parser.parse(Bin)
  end

end
